ENH: stats: Integrate library UNU.RAN in `scipy.stats` [GSoC 2021] (#14215)
scipy/scipy: doc/source/tutorial/stats/sampling.rst

.. _non-uniform-random-number-sampling:

=====================================================

> Universal Non-Uniform Random Number Sampling in SciPy

Reviewer: we should also mention the relationship to the distributions in `scipy.stats`: these already have the rvs method. The general aim is to provide a fast generation method in rv_continuous. The tools presented here allow the user to experiment with different methods that can lead to performance improvements in certain situations, e.g., if a large number of samples is needed or if particular shape parameters are used.

+ that while the same uniform rng is used, the resulting rvs will be different in general

Author: > we should also mention the relationship to the distributions in `scipy.stats`: these already have the rvs method. The general aim is to provide a fast generation method in rv_continuous. The tools presented here allow the user to experiment with different methods that can lead to performance improvements in certain situations, e.g., if a large number of samples is needed or if particular shape parameters are used.
> 
> and mention that while the same uniform rng is used, the resulting rvs will be different in general

Ah, yes! I forgot to include this. This will be definitely useful. I will update the tutorial to include this.

Author: Done.

Reviewer: Thanks for adding a clarification:

> Please note the difference between the ``rvs`` method of the distributions present in :mod:`scipy.stats` and the one provided by these generators: The general aim of these generators is to provide a fast generation method in :class:`~rv_continuous`.

I propose to change it slighly since at the moment, the new UNU.RAN generators are not used by `rv_continuous` at all. Of course, this might change once we find a good way to use them in that class. My comment above was misleading.

Please note the difference between the ``rvs`` method of the distributions present in :mod:`scipy.stats` and the one provided by these generators. In general, the aim of the generation methods implemented in :class:`~rv_continuous` is to provide generators that work reasonably well for common use cases, e.g., to generate small samples using different shape parameters or to generate large samples for a fixed shape parameter. Depending on the goal, specialized methods can lead to a substantial speedup. For example, at the cost of a rather expensive setup step, very fast generation can be achieved for large samples and a fixed shape parameter.

Author: This sounds great! I will change it this way.

Author: I have changed this according to your comment.

`np.random.SeedSequence`.

.. warning:: Use of NumPy < 1.19.0 is discouraged as it doesn't have a fast
             Cython API for generating uniform random numbers and might be
             too slow to practically rely on.

All the generators have a common ``rvs`` method that can be used to draw
samples from the given distribution.

An example of this interface is shown below:

    >>> from scipy.stats import TransformedDensityRejection
    >>> from math import exp
    >>> 
    >>> class StandardNormal:
    ...     def pdf(self, x: float) -> float:
    ...         # note that the normalization constant isn't required
    ...         return exp(-0.5 * x*x)
    ...     def dpdf(self, x: float) -> float:
    ...         return -x * exp(-0.5 * x*x)
    ... 
    >>> dist = StandardNormal()
    >>> 
    >>> urng = np.random.default_rng(0x6be00a9336fc82258fe73bfa672d3d52)
    >>> rng = TransformedDensityRejection(dist, seed=urng)

As shown in the example, we first initialize a distribution object that
contains an implementation of the methods required by the generator. In
our case, we use the :class:`~TransformedDensityRejection` method which

> requires a PDF and its derivative w.r.t. x (i.e. the variate). Note that

Reviewer: should we highlight that point even more (e.g., in bold or .. note :: ...) and point out that it applies to the other methods like dpdf, too?

Author: > should we highlight that point even more (e.g., in bold or .. note :: ...) and point out that it applies to the other methods like dpdf, too?

This is a very good point. I have made this a note now.

Random variate generation is the small field of research that deals with
algorithms to generate random variates from various distributions. It is
common to assume that a uniform random number generator is available.
This is a program that produces a sequence of independent and identically
distributed continuous U(0,1) random variates (i.e. uniform random variates
on the interval (0,1)). Of course, real-world computers can never generate
ideal random numbers and they cannot produce numbers of arbitrary precision
but state-of-the-art uniform random number generators come close to this
aim. Thus random variate generation deals with the problem of transforming
such a sequence of U(0,1) random numbers into non-uniform random variates.
These methods are universal and work in a black-box fashion.

Some methods to do that are:

* The Inversion method: When the inverse :math:`F^{-1}` of the cumulative
  distribution function is known, then random variate generation is easy.
  We just generate a uniformly U(0,1) distributed random number U and
  return :math:`X = F^{-1}(U)`. As closed form solutions for the inverse
  are rarely available, one usually needs to rely on approximations of

>   the inverse. See :class:`~NumericalInverseHermite`.

Reviewer 2: Also point out that some methods in SciPy provide the inverse.

>   the inverse. See :class:`~NumericalInverseHermite`.

Reviewer: we will soon update the code of `NumericalInverseHermite` to rely on UNU.RAN

for the second point, we could say: As closed form solutions for the inverse are rarely available, one usually needs to rely on approximations of the inverse or the implementation of special functions (we could refer to an example in scipy.special). 

we could also mention that in general, the implementation of special functions are quite slow compared to the inversion methods in UNU.RAN. I recently tried the inverse of the Gamma CDF, it is about 25x slower than polynomial inversion in UNU.RAN).

>   the inverse. See :class:`~NumericalInverseHermite`.

Author: I have changed this (in 6ef498c35571bd872fd4c35dc2a8880f034dd90c) as per your suggestion @chrisb83. Thanks!

* The Rejection Method: The rejection method, often called
  acceptance-rejection method, has been suggested by John von Neumann in
  1951 [1]_. It involves computing an upper bound to the PDF (also called the
  hat function) and using the inversion method to generate a random
  variate, say Y, from this bound. Then a uniform random number can be
  drawn between 0 to the value of the upper bound at Y. If this number
  is less than the PDF at Y, return the sample otherwise reject it. See
  :class:`~TransformedDensityRejection`.
* The Ratio-of-Uniforms Method: This is a type of acceptance-rejection
  method which is uses minimal bounding rectangles to construct the hat
  function. See :func:`~rvs_ratio_uniforms`.

> * Inversion for Discrete Distributions: The difference compared to the

Author: > Do we have something to do this? We should clarify this either saying this is not supported, if it is, how.

There are no methods to do this as of now, but I will add them shortly after this is merged. I designed this keeping in mind, we would have a wrapper for most of the methods in UNU.RAN. For now, the best thing to do is to add a line saying this is not supported.

> * Inversion for Discrete Distributions: The difference compared to the

Reviewer 2: Ok great, yes a line would be enough.

> * Inversion for Discrete Distributions: The difference compared to the

Reviewer 2: Maybe consider adding a small paragraph about Metropolis Hasting and alike? Even if you started with `some of the methods`, these are quite famous and people might wonder why we are not talking about these.

Also, I think we should clarify somewhere that this is for univariate sampling and people who want to generate multivariate samples should be made aware that each variable would then be independent (no correlation matrix or copula story here.).

> * Inversion for Discrete Distributions: The difference compared to the

Author: > Maybe consider adding a small paragraph about Metropolis Hasting and alike?

Do we want to touch on MCMC in SciPy? I can make the documentation more elaborate but that would be better to be in a follow-up. Let me know what you think.

> * Inversion for Discrete Distributions: The difference compared to the

Reviewer 2: I don't think so, but it does not mean either we should not say it exists. Optional here, I just don't want non expert to be under the impression that this is the ground truth and there is nothing else.

  U(0, 1) and probabilities are summed until the cumulative probability
  exceeds the uniform random number. The index at which this happens is
  the required random variate and is returned.

More details on these algorithms can be found in the `appendix of the UNU.RAN
user manual <http://statmath.wu.ac.at/software/unuran/doc/unuran.html#RVG>`__.

Basic concepts of the Interface
-------------------------------

Every generator needs to be set up before one can start sampling from it.
This can be done by instantiating an object of that class. Most of the
generators take a distribution object as input which contains the implementation
of required methods like PDF, CDF, etc. In addition to the distribution
object, one can also pass parameters used to set up the generator. It is also
possible to truncate the distributions using a ``domain`` parameter.  All
generators need a stream of uniform random numbers that are transformed into
random variates of the given distribution. This is done by passing a ``seed``
parameter with a NumPy BitGenerator as the uniform random number generator.
``seed`` can either be a integer, `np.random.Generator`,

> `np.random.BitGenerator`, `np.random.RandomState`, or

Reviewer 3: I disagree; leave it as is. If you are describing what the function accepts, describe it accurately and completely, which means referring to `RandomState`. When you _show_ how you want people to use it, as in examples, use `Generator`.

> `np.random.BitGenerator`, `np.random.RandomState`, or

Reviewer 2: I am closing this duplicate discussion in favour of the same here https://github.com/scipy/scipy/pull/14215#discussion_r671850541

`np.random.SeedSequence`.

.. warning:: Use of NumPy < 1.19.0 is discouraged as it doesn't have a fast
             Cython API for generating uniform random numbers and might be
             too slow for practical use.

All the generators have a common ``rvs`` method that can be used to draw
samples from the given distribution.

An example of this interface is shown below:

    >>> from scipy.stats import TransformedDensityRejection
    >>> from math import exp
    >>> 
    >>> class StandardNormal:
    ...     def pdf(self, x: float) -> float:
    ...         # note that the normalization constant isn't required
    ...         return exp(-0.5 * x*x)
    ...     def dpdf(self, x: float) -> float:
    ...         return -x * exp(-0.5 * x*x)
    ... 
    >>> dist = StandardNormal()
    >>> 

>     >>> urng = np.random.default_rng(0x6be00a9336fc82258fe73bfa672d3d52)

Author: > What is the format of this seed?

I just do:

```python
import secrets
print(hex(secrets.randbits(128)))
```

> Also normally we should not seed in example (cf all our discussions about this.).

I am aware that doing this is not the best practice. But they are very random so we can probably get away with this?

> Any strong argument that we must explicitly seed (as opposed to let the doctests use the common seed we have for all the tests) for the following examples?

I wasn't aware that doctests also set the seed for code in tutorials. In that case, I will remove the explicit seed and change the output accordingly.

>     >>> urng = np.random.default_rng(0x6be00a9336fc82258fe73bfa672d3d52)

Reviewer 2: Great thanks! Although nice use of the `secret` module.

>     >>> urng = np.random.default_rng(0x6be00a9336fc82258fe73bfa672d3d52)

Author: > Although nice use of the `secret` module.

Thank @rkern. I just stole this from https://github.com/scipy/scipy/pull/14224#pullrequestreview-682348531

>     >>> urng = np.random.default_rng(0x6be00a9336fc82258fe73bfa672d3d52)

Reviewer 2: I did not comment below, but don't forget the other occurrence of these.

>     >>> urng = np.random.default_rng(0x6be00a9336fc82258fe73bfa672d3d52)

Author: I will remove all the set `seed` in the next commit

    >>> rng = TransformedDensityRejection(dist, seed=urng)

As shown in the example, we first initialize a distribution object that
contains an implementation of the methods required by the generator. In
our case, we use the :class:`~TransformedDensityRejection` method which

> requires a PDF and its derivative w.r.t. x (i.e. the variate).

Author: Done.

.. note:: Note that the methods of the distribution (i.e. ``pdf``,
          ``dpdf``, etc) need not be vectorized. They should
          accept and return floats.

.. note:: One can also pass the SciPy distributions as arguments but it can
          be slow due to validations and expensive NumPy operations.
          Moreover, it doesn't always have all the information required

>           by some generators like derivative of PDF for TDR method. Also,

Author: Defined now. See 6ef498c35571bd872fd4c35dc2a8880f034dd90c

          most of the distributions in SciPy provide a ``rvs`` method which
          can be used instead.

In the above example, we have set up an object of the
:class:`~TransformedDensityRejection` method to sample from a
standard normal distribution. Now, we can start sampling from our
distribution by calling the ``rvs`` method:

    >>> rng.rvs()
    -0.6853213192313732

>     >>> rng.rvs((5, 3))

Author: > Instead of this, I would show some plots instead.

That's a nice idea. This is missing pretty plots. Adding those should make this more interesting.

>     >>> rng.rvs((5, 3))

Reviewer 2: To get you started I have this for instance

```python
fig, ax = plt.subplots()
x = np.linspace(distribution.ppf(0.01), distribution.ppf(0.99), 100)
# PDF
pdf = distribution.pdf(x)
ax.plot(x, pdf, "-", lw=5, alpha=0.6)
# Empirical PDF
sample = distribution.rvs(size=1000)
# discrete samples
delta = np.max(pdf) * 5e-2
ax.plot(sample[:100], -delta - delta * np.random.random(100), "+k")
ax.set_title("PDF")
ax.set_ylabel(r"$f$")
ax.set_xlabel(r"$x$")
```

    array([[ 0.24129542,  1.28709556,  0.43244247],
           [ 0.88977668,  1.22981663, -0.09820373],
           [ 1.38588147, -1.06629264,  0.54475257],
           [-0.55153436, -0.67281406,  0.01830143],
           [ 1.37709649, -1.33036774,  0.62131967]])

.. note:: Please note the difference between the `rvs` method of the
          distributions present in :mod:`scipy.stats` and the one provided
          by these generators. In general, the aim of the generation
          methods implemented in :class:`~rv_continuous` is to provide
          generators that work reasonably well for common use cases, e.g.
          to generate small samples using different shape parameters or to

>           generate large samples for a fixed shape parameter. Depending on

Reviewer 2: I am not sure I understand. Does it mean that only some values can be used to generate large samples?

Author: > Does it mean that only some values can be used to generate large samples?

I assume you are talking about the line:

```
          to generate small samples using different shape parameters or to
          generate large samples for a fixed shape parameter.
```

I think it is saying this because when we change the shape parameters, the setup step needs to be run again. Hence, it is more common to draw a small number of samples between every setup step. In contrast, when we fix the shape parameters, the setup is run only once and we can draw a large number of samples in the same time. That is what the "*common* use cases" is supposed to mean, I guess. Do you think making this explicit would make it easier for general users to grasp this?

@chrisb83 Do you agree with the above explanation?

Reviewer 2: I think yes. I am trying to read this with not to much prior (and in this case I was not sure what you meant, now it's clear you are talking about freezing the distribution) and it don't really get this otherwise.

Reviewer: ok, probably we need to add a bit more information: in random variate generation, you are interested in the fixed and the varying parameter case. 

- fixed: fix a shape parameter and draw samples => in general, this is the easier case since you can run a slow setup step once and then draw samples again and again
- varying: the shape parameter keeps changing when you draw new samples. an example is Gibbs sampling => the distribution of the next sample depends on the current one. in that case, you cannot use an expensive setup. or consider simulating the initial wealth of 100k agents, each drawn from a distribution with a different mean

so choosing the right method strongly depends on the situation...

Reviewer: since this is an important topic, we should explain these concepts better. in the UNU.RAN doc, the authors give an indication of the setup and sampling speed, e.g.,

Automatic ratio of uniforms method: Set-up: slow, Sampling: fast
Simple ratio of uniforms method: Set-up: fast, Sampling: slow

I think we should create a table in the tutorial as follows:

| Method | Required inputs | Setup speed | Sampling Speed |
| SimpleRatioUniforms | pdf, ... | fast | slow |

@tirthasheshpatel I can draft a short introduction to explain these concepts better.

Author: > since this is an important topic, we should explain these concepts better.

I agree.

> I think we should create a table in the tutorial as follows:

I think a summary table like the one you have made would be very useful. I have also mentioned this information in the tutorial of methods (see `sampling_tdr`, `sampling_dau`).

> I can draft a short introduction to explain these concepts better.

That would be a lot of help @chrisb83, thanks!

Author: I have added a simple table @chrisb83 (see ea5e083219a3223b33b2aca3a1c94cb8c9659622). Let me know what you think about it.

Reviewer: Great! I would just suggest to create separate tables for continuous and discrete distributions

Reviewer 2: This is really helpful, thanks!

Author: @chrisb83 I have separated discrete and continuous methods now.

Reviewer 2: Feel free to resolve this one. On my side it's covered.

          the goal, specialized methods can lead to a substantial speedup.
          For example, at the cost of a rather expensive setup step, very
          fast generation can be achieved for large samples and a fixed
          shape parameter. Also, even if the same URNG (``seed``) is used,

>           the resulting rvs will be different in general.

Reviewer 2: ```suggestion
          the resulting ``rvs`` will be different in general.
```
Why is that the case? It should say, just a line.

Author: > Why is that the case?

This difference exists between the `rvs` method of distributions in SciPy and the `rvs` method of UNU.RAN generators. It is because UNU.RAN transforms uniform random numbers differently than SciPy of NumPy. Is this not clear enough? Should I make this more explicit?

Reviewer 2: What I don't get with this is: If you say that having the same `seed=urng` does not output the same `rvs`, then why do we have this parameter?

Author: > If you say that having the same `seed=urng` does not output the same `rvs`, then why do we have this parameter?

For instance, if we set `seed=urng`, the random number stream will match for the same method (say, `TransformedDensityRejection`).  But the rvs stream of SciPy's `norm` and UNU.RAN's `TransformedDensityRejection` would be the same even for the same `seed`/`random_state`:

```python
>>> from scipy.stats import norm, TransformedDensityRejection
>>> 
>>> class StdNorm:
...     def pdf(self, x):
...         return np.exp(-0.5 * x*x)
...     def dpdf(self, x):
...         return -x * self.pdf(x)
... 
>>> dist = StdNorm()
>>> rng = TransformedDensityRejection(dist, seed=123)
>>> rng.rvs()
0.474548717355228
>>> norm.rvs(random_state=123)
-1.0856306033005612
```

Moreover, different methods also produce different streams of RVs. So, for example, for the same `seed`, the output of `TransformedDensityRejection` would differ from `NumericalInverseHermite`. Hope this helps!

Reviewer 2: Ah ok, I thought you were talking about UNU.RAN only here. The `URNG` was misleading I think.

Author: I have made this clearer now. (see 3dcd2689a46ffdac2ff57a949b46d9a165c55d83)

>           shape parameter. Also, even if the same URNG (``seed``) is used,
>           the resulting rvs will be different in general.

Reviewer 2: I propose something along this line
```suggestion
          shape parameter. Also, even if the same ``seed`` is used,
          the resulting `rvs` will be different between the generators.
```

.. note:: One can also pass the SciPy distributions as arguments but it can
          be slow due to validations and expensive NumPy operations.
          Moreover, it doesn't always have all the information required
          by some generators like derivative of PDF for TDR method. Also,
          most of the distributions in SciPy provide a ``rvs`` method which
          can be used instead.

In the above example, we have set up an object of the
:class:`~TransformedDensityRejection` method to sample from a
standard normal distribution. Now, we can start sampling from our
distribution by calling the ``rvs`` method:

    >>> rng.rvs()
    -0.6853213192313732
    >>> rng.rvs((5, 3))
    array([[ 0.24129542,  1.28709556,  0.43244247],
           [ 0.88977668,  1.22981663, -0.09820373],
           [ 1.38588147, -1.06629264,  0.54475257],
           [-0.55153436, -0.67281406,  0.01830143],
           [ 1.37709649, -1.33036774,  0.62131967]])

.. note:: Please note the difference between the `rvs` method of the
          distributions present in :mod:`scipy.stats` and the one provided
          by these generators. Even if the same URNG (``seed``) is used,

>           the resulting rvs will be different in general: the implementation
>           of `rvs` in :class:`~rv_continuous` usually relies on the NumPy
>           module `random` for well-known distributions (e.g., for the normal

Reviewer 2: ```suggestion
          the resulting ``rvs`` will be different in general: the implementation
          of `rvs` in :class:`~rv_continuous` usually relies on the NumPy
          module `np.random` for well-known distributions (e.g., for the normal
```

          distribution, the beta distribution) and transformations of other
          distributions (e.g., normal inverse Gaussian `norminvgauss` and the

>           lognormal distribution). If no specific  method is implemented,

Reviewer 2: ```suggestion
          lognormal distribution). If no specific method is implemented,
```

standard normal distribution. Now, we can start sampling from our
distribution by calling the ``rvs`` method:

    >>> rng.rvs()
    -1.526829048388144
    >>> rng.rvs((5, 3))
    array([[ 2.06206883,  0.15205036,  1.11587367],
           [-0.30775562,  0.29879802, -0.61858268],
           [-1.01049115,  0.78853694, -0.23060766],
           [-0.60954752,  0.29071797, -0.57167182],
           [ 0.9331694 , -0.95605208,  1.72195199]])

.. plot::

    >>> import matplotlib.pyplot as plt
    >>> fig, ax = plt.subplots()
    >>> x = np.linspace(norm.ppf(0.01), norm.ppf(0.99), 100)
    >>> # PDF
    >>> pdf = norm.pdf(x)
    >>> ax.plot(x, pdf, "-", lw=5, alpha=0.6)
    >>> # Empirical PDF
    >>> sample = rng.rvs(size=1000)
    >>> # discrete samples
    >>> delta = np.max(pdf) * 5e-2
    >>> ax.plot(sample[:100], -delta - delta * np.random.random(100), "+k")
    >>> ax.set_title("PDF")
    >>> ax.set_ylabel(r"$f$")
    >>> ax.set_xlabel(r"$x$")
    >>> plt.show()

.. note:: Please note the difference between the `rvs` method of the
          distributions present in :mod:`scipy.stats` and the one provided
          by these generators. Even if the same URNG (``seed``) is used,

>           the resulting ``rvs`` will be different in general: the implementation

Reviewer 4: This reads like the `rvs` methods aren't stable even for the same seed. I think it would be clearer to write that anything passed through UNU.RAN must be considered independent and that it will produce different results (than potential scipy-equivalents) for any seed.

Author: > I think it would be clearer to write that anything passed through UNU.RAN must be considered independent and that it will produce different results (than potential scipy-equivalents) for any seed.

Thanks for the suggestion! I have changed this now to avoid any ambiguity. Let me know if it looks good.

          of `rvs` in :class:`~rv_continuous` usually relies on the NumPy
          module `np.random` for well-known distributions (e.g., for the normal
          distribution, the beta distribution) and transformations of other
          distributions (e.g., normal inverse Gaussian `norminvgauss` and the
          lognormal distribution). If no specific method is implemented,
          `rv_continuous` defaults to a numerical inversion method of the CDF
          that is very slow. As UNU.RAN transforms uniform random numbers
          differently than SciPy or NumPy, the resulting stream of RVs is
          different even for the same stream of uniform random numbers. For
          example, the random number stream of SciPy's ``~norm`` and UNU.RAN's
          :class:`~TransformedDensityRejection` would not be the same even for
          the same seed/random_state:

          >>> from scipy.stats import norm, TransformedDensityRejection
          >>> 
          >>> dist = StandardNormal()

>           >>> rng = TransformedDensityRejection(dist, seed=123)

Reviewer 2: Fixed seed

Author: Oops, forgot this one.

Author: This is to demonstrate that even for the same seed, RVS of numpy distributions and UNU.RAN can differ.

Reviewer 2: Ok but then I would use something less digest like you have wit `0x7344aca86f1dafd8820be998b639551c`

Reviewer 2: I would be fine with truncating a bit this value like so `0x7344aca86f1` for instance. But definitely not 123 or something like this please. You will always have people copy pasting code and changing the bare minimum...

Author: I have removed the explicit seed and just made a copy. Does it look good now?

Reviewer 2: I personally prefer this approach, thanks.

    -1.526829048388144
    >>> rng.rvs((5, 3))
    array([[ 2.06206883,  0.15205036,  1.11587367],
           [-0.30775562,  0.29879802, -0.61858268],
           [-1.01049115,  0.78853694, -0.23060766],
           [-0.60954752,  0.29071797, -0.57167182],
           [ 0.9331694 , -0.95605208,  1.72195199]])

.. plot::

    >>> import matplotlib.pyplot as plt
    >>> from scipy.stats import norm, TransformedDensityRejection
    >>> from math import exp
    >>> 
    >>> class StandardNormal:
    ...     def pdf(self, x: float) -> float:
    ...         # note that the normalization constant isn't required
    ...         return exp(-0.5 * x*x)
    ...     def dpdf(self, x: float) -> float:
    ...         return -x * exp(-0.5 * x*x)
    ... 
    >>> 
    >>> dist = StandardNormal()
    >>> urng = np.random.default_rng()
    >>> rng = TransformedDensityRejection(dist, seed=urng)
    >>> fig, ax = plt.subplots()
    >>> x = np.linspace(norm.ppf(0.01), norm.ppf(0.99), 100)
    >>> # PDF
    >>> pdf = norm.pdf(x)
    >>> ax.plot(x, pdf, "-", lw=5, alpha=0.6)
    >>> # Empirical PDF
    >>> sample = rng.rvs(size=1000)
    >>> # discrete samples
    >>> delta = np.max(pdf) * 5e-2

>     >>> ax.plot(sample[:100], -delta - delta * np.random.random(100), "+k")

Reviewer: I just looked at the generated plot and I do not understand how it helps us to understand that the samples are normally distributed. wouldn't a simple histogram be easier to understand?

Author: > wouldn't a simple histogram be easier to understand?

Yes, I think so. I have added some histogram plots in tuts and docs of both methods. Let's see how they look.

Reviewer: That looks very nice!

>     >>> ax.set_title("PDF")

Reviewer: title is a bit unclear: is the purpose of the plot just to show the pdf?

Author: changed.

    >>> ax.set_ylabel(r"$f$")
    >>> ax.set_xlabel(r"$x$")
    >>> plt.show()

.. note:: Please note the difference between the `rvs` method of the
          distributions present in :mod:`scipy.stats` and the one provided

>           by these generators. UNU.RAN generators must be considered

Reviewer: i think we should add a disclaimer that the streams will be different *in general*, e.g., if SciPy uses an inversion of the cdf and we use a UNU.RAN inversion method, the differences of the rvs will probably be very minor. maybe:

> in a sense that they will in general produce a different stream of random numbers than the one produced by the equivalent distribution in :mod:`scipy.stats` for any seed

Author: done.

samplers to sample random variates from a wide variety of univariate
continuous and discrete distributions. Implementations of a fast C library
called `UNU.RAN <http://statmath.wu.ac.at/software/unuran/>`__ are used
for speed and performance. Please look at
`UNU.RAN's documentation <http://statmath.wu.ac.at/software/unuran/doc/unuran.html>`__
for an in-depth explanation of these methods. It is heavily referred
for writing this tutorial and the documentation of all the samplers.

Introduction
------------

Random variate generation is the small field of research that deals with
algorithms to generate random variates from various distributions. It is
common to assume that a uniform random number generator is available.
This is a program that produces a sequence of independent and identically
distributed continuous U(0,1) random variates (i.e. uniform random variates
on the interval (0,1)). Of course, real-world computers can never generate
ideal random numbers and they cannot produce numbers of arbitrary precision
but state-of-the-art uniform random number generators come close to this
aim. Thus random variate generation deals with the problem of transforming

> such a sequence of U(0,1) random numbers into non-uniform random variates.

Author: I have mentioned it in section "*Basic Concepts of the Interface*":

> [possible to truncate the distributions using a ``domain`` parameter.]  **All
> generators need a stream of uniform random numbers that are transformed into
> random variates of the given distribution. This is done by passing a ``seed``
> parameter with a NumPy BitGenerator as the uniform random number generator.**

> such a sequence of U(0,1) random numbers into non-uniform random variates.

Reviewer: Ok, let's keep it as it is

Some methods to do that are:

* The Inversion method: When the inverse :math:`F^{-1}` of the cumulative
  distribution function is known, then random variate generation is easy.
  We just generate a uniformly U(0,1) distributed random number U and

>   return :math:`X = F^{-1}(U)`. See

Reviewer: Let's add something like: As closed form solutions for the inverse are rarely available, one usually needs to rely on approximations of the inverse. The followings approaches are implemented in SciPy:

Author: Done.

  :class:`~NumericalInverseHermite`.
* The Rejection Method: The rejection method, often called
  acceptance-rejection method, has been suggested by John von Neumann in

>   1951. It involves computing an upper bound to the PDF (also called the

Author: Done.

  variate, say Y, from this bound. Then a uniform random number can be
  drawn between 0 to the value of the upper bound at Y. If this number
  is less than the PDF at Y, return the sample otherwise reject it. See
  :class:`~TransformedDensityRejection`.
* The Ratio-of-Uniforms Method: This is a type of acceptance-rejection
  method which is uses minimal bounding rectangles to construct the hat
  function. See :func:`~rvs_ratio_uniforms`.
* Inversion for Discrete Distributions: The difference compared to the
  continuous case is that :math:`F` is now a step-function. To realize
  this in a computer, a search algorithm is used, the simplest of which
  is *sequential search*. A uniform random number is generated from
  U(0, 1) and probabilities are summed until the cumulative probability
  exceeds the uniform random number. The index at which this happens is
  the required random variate and is returned.

More details on these algorithms can be found in the `appendix of the UNU.RAN
user manual <http://statmath.wu.ac.at/software/unuran/doc/unuran.html#RVG>`__.

Basic concepts of the Interface
-------------------------------

> Every sampler needs to be set up before one can start sampling from it.

Reviewer: Should we say generator instead of sampler? The former term is used in UNU.RAN

Author: > Should we say generator instead of sampler?

I was also confused between the two naming conventions. So, I have renamed "sampler" to "generator" to follow one common convention.

.. note:: Please note the difference between the `rvs` method of the
          distributions present in :mod:`scipy.stats` and the one provided
          by these generators. Even if the same URNG (``seed``) is used,
          the resulting rvs will be different in general: the implementation
          of `rvs` in :class:`~rv_continuous` usually relies on the NumPy
          module `random` for well-known distributions (e.g., for the normal
          distribution, the beta distribution) and transformations of other
          distributions (e.g., normal inverse Gaussian `norminvgauss` and the
          lognormal distribution). If no specific  method is implemented,
          `rv_continuous` defaults to a numerical inversion method of the CDF
          that is very slow. As described in this tutorial, UNU.RAN implements

>           various approaches that will generally produce different samples.

Author: This elucidates some of the subtle concepts in the tutorial. I went through the diff and it looks good to me. Thanks for working on this, @chrisb83! I will add the table next Monday.

How does the new description sound to you, @tupui? (It's based on https://github.com/scipy/scipy/pull/14215#discussion_r672824528)

Reviewer 2: Sound better for me as well, thanks! Just a few suggestions for rendering.
